trussworks/react-uswds
# feat: Normalize component structure and exports

# Summary

Everything noted below is under the pretense of NOT breaking existing usage patterns. All changes are intended to be additions, with JSDoc deprecation comments used where appropriate.

- ensure all compositing components have a `[Component]Base` component that focuses on root element rendering, if applicable
- create components for all currently inlined elements that require logic (ex: className composition)
  - ensure all new components have basic tests (render, className) in addition to moving tests from parent component test that are within its area of concern
- ensure component type exports
  - normalize exported `Base[Component]Props` as type holding custom props
  - normalize exported `[Component]Props` as `React.ComponentPropsWithRef<typeof [Component]>`
  - normalize `[Component]Ref` as `React.ComponentRef<typeof [Component]>`
    - Parent components that pass on their forwardRef to these components can use this ref type for `React.ForwardRefRenderFunction` or wherever else appropriate as a reliable alias
  - typeof shortcutting to ensure changes to ref type or prop types in forwardRefFn are always reflected
  - Use `React.[Component]PropsWith[out]Ref` for props that are intended to be directly forwarded to child component
    - This type automatically handles the ref typing so that forwardRefs are compatible (ref from `JSX.IntrinsicElements` is type-incompatible)
- all components are forwardRef, always passing the forwardRef to most appropriate child component
- compositing components expose child component prop bags that are passed on (which will now include ref)
- JSDoc deprecation marking of certain props in favor of new ones
  - `divProps` -> `headerProps`
  - dedicated prop that was used as `children` -> `children`
- other minor enhancements
  - normalize `isX` name format for booleans (adding deprecation markers where appropriate)
  - ex: allowing boolean props to be `undefined` for falseyness
  - exposing props for emitting on certain events to make components less of a black box (ex: emit currently open accordion items when changed via `onChange` prop on `Accordion`)
- export any other types or functions that can be useful

## Related Issues or PRs

#2667 
#2666
#2587 
#2232 
#2599 
#2000 


## How To Test

- Use this branch for your local project's version of react-uswds
- Ensure all exports are available
- If possible, verify that existing code using react-uswds is unbroken

### Screenshots (optional)
